rfcs/0018-glides-trainsheet-events.md

```diff
@@ -0,0 +1,571 @@
+- `arrivalTime` (Time, optional): if present, the new time that the train is expected to arrive at the given location. If `location` is not specified, this is the time the vehicle is arriving at the final location of the trip.
+- `departureTime` (Time, optional): if present, the new time that the train is expected to depart the given location. If `location` is not specified, this is the time the vehicle is departing the first location of the trip.
+- `consist` (array of Optional Car, optional): the cars assigned to perform this trip. If a car is `null`, then no car is currently assigned to that position in the consist. The front car should be listed first.
+- `operators` (array of Optional Operator, optional): the list of operators assigned to perform this trip. If an operator is `null`, then no operator is assigned to the car in that position in the consist. If the operator is "unmodified" then the operator is the value from the schedule or the most-recent Trip updated message. 
+
+*Notes*
+- `consist` and `operators` MUST be the same length; if they are not, the behavior of consumers is undefined.
+
+# Reference-level explanation
+## Effects on RTR
+These inputs should be similar to what RTR is already receiving from OCS.
+- Trip added: `TSCH NEW`
+- Trip dropped: `TSCH DEL` with delete status 1
+- Trip restored: `TSCH DEL` with delete status 0
+- Trip updated: `TSCH CON` (set consist), `TSCH OFF` (change departure time)
+
+One key difference is that the trainsheets do not record the destination or route for trips, so added trips currently only include one end of the trip. RTR should assume a default pattern for these trips.
+
+## Long-term storage / querying
+For future use, events will be archived to S3, using Kinesis Firehose. They can either be sent to LAMP's folder, or a new folder. As the stream contains potential PII (operator badge information) any long-term storage (S3, database) must be encrypted at rest.
+
+### Schema
+All events should have their schema documented in the [mbta/schemas](https://github.com/mbta/schemas) GitHub repository.
+
+### Compatibility
+Changes to the events can happen in one of two ways:
+- backwards-compatible changes (such as adding a new field with a default value) where old consumers can still interpret new events can be made while keeping the same event type. 
+- incompatible-changes require creating a new event type. Some options as described in [RFC 4](https://github.com/mbta/technology-docs/blob/main/rfcs/accepted/0004-socket-proxy-ocs-cloudevents.md) are `<type>.<date>` (adding the date of the new event) or `<type>.v2` (creating a new version). Each version should be documented separately in the [mbta/schemas](https://github.com/mbta/schemas) GitHub repository.
+
+Consumers MUST ignore event types that they do not understand.
```

Comment:
This [OUGHT TO](https://www.rfc-editor.org/rfc/rfc6919#section-4) explicitly state that consumers should ignore extra fields in events which they do understand. It's loosely implied by the above declaration that adding an extra field is backwards-compatible, but that's only backwards compatible if consumers are forwards compatible.

Reply:
84a9be343fd03649f330eee0f387d22b431f2681